amethyst_tiles/src/pass.rs

```diff
@@ -0,0 +1,444 @@
+            change: Default::default(),
+        }))
+    }
+}
+
+/// `RenderGroup` providing culling, drawing and transparency functionality for 3D `TileMap` components.
+#[derive(Derivative)]
+#[derivative(Debug(bound = ""))]
+pub struct DrawTiles2D<
+    B: Backend,
+    T: Tile,
+    E: CoordinateEncoder,
+    Z: DrawTiles2DBounds = DrawTiles2DBoundsDefault,
+> {
+    pipeline: B::GraphicsPipeline,
+    pipeline_layout: B::PipelineLayout,
+    textures: TextureSub<B>,
+    vertex: DynamicVertexBuffer<B, TileArgs>,
+    sprites: OrderedOneLevelBatch<TextureId, TileArgs>,
+    change: util::ChangeDetection,
+
+    env: DynamicUniform<B, TileMapArgs>,
+
+    #[derivative(Debug = "ignore")]
+    _marker: PhantomData<(T, E, Z)>,
+}
+
+impl<B: Backend, T: Tile, E: CoordinateEncoder, Z: DrawTiles2DBounds> RenderGroup<B, World>
+    for DrawTiles2D<B, T, E, Z>
+{
+    #[allow(clippy::cast_precision_loss)]
+    fn prepare(
+        &mut self,
+        factory: &Factory<B>,
+        _queue: QueueId,
+        index: usize,
+        _subpass: hal::pass::Subpass<'_, B>,
+        world: &World,
+    ) -> PrepareResult {
+        #[cfg(feature = "profiler")]
+        profile_scope!("prepare");
+
+        let mut changed = false;
+        let (sprite_sheet_storage, tex_storage, hiddens, tile_maps, transforms) =
+            <(
+                Read<'_, AssetStorage<SpriteSheet>>,
+                Read<'_, AssetStorage<Texture>>,
+                ReadStorage<'_, Hidden>,
+                ReadStorage<'_, TileMap<T, E>>,
+                ReadStorage<'_, Transform>,
+            )>::fetch(world);
+
+        let sprites_ref = &mut self.sprites;
+        let textures_ref = &mut self.textures;
+
+        sprites_ref.swap_clear();
+
+        let CameraGatherer { projview, .. } = CameraGatherer::gather(world);
+
+        let mut tilemap_args = TileMapArgs {
+            proj: projview.proj,
+            view: projview.view,
+            map_coordinate_transform: Default::default(),
+            map_transform: Default::default(),
+            sprite_dimensions: Default::default(),
+        };
+
+        if let Some((tile_map, _, transform)) =
+            (&tile_maps, !&hiddens, transforms.maybe()).join().next()
+        {
+            let map_coordinate_transform: [[f32; 4]; 4] = (*tile_map.transform()).into();
+
+            let map_transform: [[f32; 4]; 4] = if let Some(transform) = transform {
+                (*transform.global_matrix()).into()
+            } else {
+                Matrix4::identity().into()
+            };
+
+            tilemap_args.map_coordinate_transform = map_coordinate_transform.into();
+            tilemap_args.map_transform = map_transform.into();
+            tilemap_args.sprite_dimensions = [
+                tile_map.tile_dimensions().x as f32,
+                tile_map.tile_dimensions().y as f32,
+            ]
+            .into();
+
+            let region = Z::bounds(tile_map, world);
+            region
+                .iter()
```

Comment:
This iteration order assumes certain depth order and this isn't exactly documented. What if you look at the tilemap from the back side? I believe the semitransparent layers would be rendered incorrectly.

I don't even want to begin thinking about multiple multilayer tilemaps passing through each other :P
Maybe we should just document some cases as "undefined behaviour" (or at least, undefined draw order) for now?

Reply:
Yes. I will document that tilemaps must be viewed from Z-up currently.